src/main/java/org/opensearch/knn/index/KNNWeight.java

```diff
@@ -117,12 +124,34 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
             final KNNQueryResult[] results;
             KNNCounter.GRAPH_QUERY_REQUESTS.increment();
 
+            // We need to first get index allocation
+            NativeMemoryAllocation indexAllocation = null;
+
+            try {
+                indexAllocation = nativeMemoryCacheManager.get(
+                        new NativeMemoryEntryContext.IndexEntryContext(
+                                indexPath.toString(),
+                                NativeMemoryLoadStrategy.IndexLoadStrategy.INSTANCE,
+                                ImmutableMap.of(SPACE_TYPE, spaceType.getValue()),
+                                knnQuery.getIndexName()
+                        ), true);
+            } catch (ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+
+            // Now that we have the allocation, we need to readLock it
+            indexAllocation.readLock();
+
             try {
-                results = knnIndexCache.queryIndex(indexPath.toString(), knnQuery.getIndexName(),
-                        spaceType, knnQuery.getQueryVector(), knnQuery.getK(), knnEngine.getName());
-            } catch (Exception ex) {
-                KNNCounter.GRAPH_QUERY_ERRORS.increment();
-                throw ex;
+                if (indexAllocation.isClosed()) {
+                    throw new RuntimeException("Index has already been closed");
+                }
+
+                results = JNIService.queryIndex(indexAllocation.getPointer(), knnQuery.getQueryVector(), knnQuery.getK(), knnEngine.getName());
```

Comment:
Will this raise exception?

Reply:
It can potentially. If there is an issue in JNI, an exception will propagate up